src/Ocean/Model/HydrostaticBoussinesqModel.jl

```diff
@@ -0,0 +1,515 @@
+  @vars begin
+    u::SVector{2, T}
+    η::T # real a 2-D variable TODO: should be 2D
+    θ::T
+  end
+end
+
+# If this order is changed check update_aux!
+function vars_aux(m::HBModel, T)
+  @vars begin
+    w::T
+    pkin_reverse::T # ∫(-αᵀ θ) # TODO: remove me after better integral interface
+    w_reverse::T               # TODO: remove me after better integral interface
+    pkin::T         # ∫(-αᵀ θ)
+    wz0::T          # w at z=0
+    θʳ::T           # SST given    # TODO: Should be 2D
+    f::T            # coriolis
+    τ::T            # wind stress  # TODO: Should be 2D
+    # κ::SMatrix{3, 3, T, 9} # diffusivity tensor (for convective adjustment)
+    κᶻ::T
+
+    # diagnostics (should be ported elsewhere eventually)
+    div2D::T
+    θu::T
+    θv::T
+    θw::T
+    ∂θ∂z::T
+  end
+end
+
+function vars_gradient(m::HBModel, T)
+  @vars begin
+    u::SVector{2, T}
+    θ::T
+  end
+end
+
+function vars_diffusive(m::HBModel, T)
+  @vars begin
+    ν∇u::SMatrix{3, 2, T, 6}
+    κ∇θ::SVector{3, T}
+  end
+end
+
+function vars_integrals(m::HBModel, T)
+  @vars begin
+    ∇hu::T
+    αᵀθ::T
+  end
+end
+
+@inline function flux_nondiffusive!(m::HBModel, F::Grad, Q::Vars,
+                                    A::Vars, t::Real)
+  @inbounds begin
+    u = Q.u # Horizontal components of velocity
+    η = Q.η
+    θ = Q.θ
+    w = A.w   # vertical velocity
+    pkin = A.pkin
+    v = @SVector [u[1], u[2], w]
+    Ih = @SMatrix [ 1 -0;
+                   -0  1;
+                   -0 -0]
+
+    # ∇ • (u θ)
+    F.θ += v * θ
+
+    # ∇h • (g η)
+    F.u += grav * η * Ih
+
+    # ∇h • (- ∫(αᵀ θ))
+    F.u += pkin * Ih
+
+    # ∇h • (v ⊗ u)
+    # F.u += v * u'
+
+  end
+
+  return nothing
+end
+
+@inline wavespeed(m::HBModel, n⁻, _...) = abs(SVector(m.c₁, m.c₂, m.c₃)' * n⁻)
+
+# We want not have jump penalties on η (since not a flux variable)
+function update_penalty!(::Rusanov, ::HBModel, n⁻, λ, ΔQ::Vars,
+                         Q⁻, Q⁺, A⁻, A⁺, t)
+  ΔQ.η = -0
+
+  θ⁻ = Q⁻.θ
+  u⁻ = Q⁻.u
+  w⁻ = A⁻.w
+  @inbounds v⁻ = @SVector [u⁻[1], u⁻[2], w⁻]
+  n̂_v⁻ = n⁻∘v⁻
+
+  θ⁺ = Q⁺.θ
+  u⁺ = Q⁺.u
+  w⁺ = A⁺.w
+  @inbounds v⁺ = @SVector [u⁺[1], u⁺[2], w⁺]
+  n̂_v⁺ = n⁻∘v⁺
+
+  # max velocity
+  # n̂∘v = (abs(n̂∘v⁺) > abs(n̂∘v⁻) ? n̂∘v⁺ : n̂∘v⁻
+
+  # average velocity
+  n̂_v = (n̂_v⁻ + n̂_v⁺) / 2
+
+  # ΔQ.θ = ((n̂_v > 0) ? 1 : -1) * (n̂_v⁻ * θ⁻ - n̂_v⁺ * θ⁺)
+  # ΔQ.θ = abs(n̂_v⁻) * θ⁻ - abs(n̂_v⁺) * θ⁺
+
+  return nothing
+end
+
+@inline function flux_diffusive!(m::HBModel, F::Grad, Q::Vars, D::Vars,
+                                 A::Vars, t::Real)
+  F.u -= D.ν∇u
+  F.θ -= D.κ∇θ
+
+  return nothing
+end
+
+@inline function gradvariables!(m::HBModel, G::Vars, Q::Vars, A, t)
+  G.u = Q.u
+  G.θ = Q.θ
+
+  return nothing
+end
+
+@inline function diffusive!(m::HBModel, D::Vars, G::Grad, Q::Vars,
+                            A::Vars, t)
+  ν = Diagonal(@SVector [m.νʰ, m.νʰ, m.νᶻ])
+  D.ν∇u = ν * G.u
+
+  κ = Diagonal(@SVector [m.κʰ, m.κʰ, A.κᶻ])
```

Comment:
If this scaling was moved to `flux_diffusive!` could we undo PR #543?

Since the computation of `κᶻ` in [`g!`] just needs `D.κ∇θ[3]` and this would be available in `flux_diffusive!` and known to be consistent with the current state.

Just a thought since #543 still makes me nervous since `diffstate` and `state` are not guaranteed to be consistent at this point (I just see some really nasty bugs arising from that).

[`g!`]: https://github.com/climate-machine/CLIMA/blob/22df165636739cae7ed064dcaa12635c9498b40e/src/Ocean/Model/HydrostaticBoussinesqModel.jl#L285-L295

Reply:
Yes, it's easy to do that and then add an `update_aux_diffusive!` that uses the gradients calculated at the same time step.

I'll switch to just storing `∇θ` in the diffusive state and then save the diffusivity scaling for `flux_diffusive!`

Reply:
Yes, it's easy to do that and then add an `update_aux_diffusive!` that uses the gradients calculated at the same time step.

I'll switch to just storing `∇θ` in the diffusive state and then save the diffusivity scaling for `flux_diffusive!`

Reply:
@jkozdon what's your plan for making this change? should we merge this branch as is and then fix it in a separate pull request that undoes #543 ? Or do you want to fix first and then merge this in?

Reply:
Either way is fine with me. Whatever works best for you.

Reply:
Okay, I'll handle this afterwards